hack/cluster-clean.sh

```diff
@@ -25,114 +25,134 @@ source hack/common.sh
+        arr=($p)
+        name="${arr[0]}"
+        ns="${arr[1]}"
+        _kubectl patch vmsnapshots $name -n $ns --type=json -p '[{ "op": "remove", "path": "/metadata/finalizers" }]'
     done
-done
 
-# Not namespaced resources
-for label in ${labels[@]}; do
-    _kubectl delete validatingwebhookconfiguration -l ${label}
-    _kubectl delete pv -l ${label}
-    _kubectl delete clusterrolebinding -l ${label}
-    _kubectl delete clusterroles -l ${label}
-    _kubectl delete customresourcedefinitions -l ${label}
+    kubectl get vmsnapshotcontents --all-namespaces -o=custom-columns=NAME:.metadata.name,NAMESPACE:.metadata.namespace,FINALIZERS:.metadata.finalizers --no-headers | grep vmsnapshotcontent-protection | while read p; do
+        arr=($p)
+        name="${arr[0]}"
+        ns="${arr[1]}"
+        _kubectl patch vmsnapshotcontents $name -n $ns --type=json -p '[{ "op": "remove", "path": "/metadata/finalizers" }]'
+    done
 
-    # W/A for https://github.com/kubernetes/kubernetes/issues/65818
-    _kubectl delete apiservices -l ${label} --wait=false
+    # Remove finalizers from all running vmis, to not block the cleanup
+    _kubectl get vmis --all-namespaces -o=custom-columns=NAME:.metadata.name,NAMESPACE:.metadata.namespace,FINALIZERS:.metadata.finalizers --no-headers | grep foregroundDeleteVirtualMachine | while read p; do
+        arr=($p)
+        name="${arr[0]}"
+        ns="${arr[1]}"
+        _kubectl patch vmi $name -n $ns --type=json -p '[{ "op": "remove", "path": "/metadata/finalizers" }]'
+    done
 
-    _kubectl get apiservices -l ${label} -o=custom-columns=NAME:.metadata.name,FINALIZERS:.metadata.finalizers --no-headers | grep foregroundDeletion | while read p; do
+    _kubectl get vms --all-namespaces -o=custom-columns=NAME:.metadata.name,NAMESPACE:.metadata.namespace,FINALIZERS:.metadata.finalizers --no-headers | grep -e foregroundDeleteVirtualMachine -e orphan -e snapshot-source-protection | while read p; do
         arr=($p)
         name="${arr[0]}"
-        _kubectl -n ${i} patch apiservices $name --type=json -p '[{ "op": "remove", "path": "/metadata/finalizers" }]'
+        ns="${arr[1]}"
+        _kubectl patch vm $name -n $ns --type=json -p '[{ "op": "remove", "path": "/metadata/finalizers" }]'
     done
-done
-
-for i in ${managed_namespaces[@]}; do
-    if [ -n "$(_kubectl get ns | grep "${i} ")" ]; then
-        echo "Clean ${i} namespace"
-        _kubectl delete ns ${i}
-
-        start_time=0
-        sample=10
-        timeout=120
-        echo "Waiting for ${i} namespace to disappear ..."
-        while [ -n "$(_kubectl get ns | grep "${i} ")" ]; do
-            sleep $sample
-            start_time=$((current_time + sample))
-            if [[ $current_time -gt $timeout ]]; then
-                exit 1
-            fi
+}
+
+function delete_resources() {
+    managed_namespaces=("$@")
```

Comment:
Maybe we could define variables in the function scope as `local` wdyt? Also we could use `$1` instead of `$@`, maybe at some point the function can accept more parameters.

Reply:
about the `$1`
I am not sure its possible
did as suggested here for passing array
https://askubuntu.com/questions/674333/how-to-pass-an-array-as-function-argument

I prefer to keep it this way if possible please,
in case the future will need more variables it can do something else

Reply:
Sure makes sense :+1: